### PR TITLE
TT-4699 Feature/analytics serialization

### PR DESCRIPTION
**Pull request by sredxny**
## Description
Added support for protobuf as serialization method of the analytics. In order to enable it is required that the user configure protobuf as encoding method in the gateway config file. Currently is not required a configuration in tyk-pump side due that it fallbacks in msgpack but first it tries to decode it using protobuf.

## Related Issue
[TT-4699](https://tyktech.atlassian.net/browse/TT-4699)

## Motivation and Context
- Be consistent in how we encode/decode analytics across all the tyk components
- Improve the speed
- Reduce some bytes while transmiting data in the network

## How This Has Been Tested
This should be tested using the changes done in gateway side. PR:
The steps are as next:

- In the gateway set as encoding method: protobuf. In order to do so, in the config file add a new config option -> `analytics_config.serializer_type: "protobuf"`
- Add a new API (it can be a simple keyless api)
- Send some traffic to the gateway
- Check redis, there should be some records with the suffix: "_protobuf"
- Turn on the pump and the data should be processed: records should not exist anymore in redis and should be in the pump configured

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

**Comment by sredxny** [comment]
@tbuchaillot @matiasinsaurralde wonddering if we should remove all the implementation of gotiny (which makes sense to me) due that it cannot be used in productions environments

File: main.go [hunk 1]
```diff
@@ -243,6 +227,34 @@ func StartPurgeLoop(wg *sync.WaitGroup, ctx context.Context, secInterval int, ch
...
+		err := serializerMethod.Decode([]byte(v.(string)), &decoded)
+
+		log.WithFields(logrus.Fields{
+			"prefix": mainPrefix,
+		}).Debug("Decoded Record: ", decoded)
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"prefix":       mainPrefix,
+				"analytic_key": analyticsKeyName,
+			}).Error("Couldn't unmarshal analytics data:", err)
```
Comment (matiasinsaurralde) [inline main.go:244]: Perhaps just add `continue` after line 244 and remove the block on 245.

File: serializer/serializer_test.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		t.Run(tc.testName, func(t *testing.T) {
+			record := analytics.AnalyticsRecord{
+				APIID: "api_1",
+				OrgID: "org_1",
+			}
+
+			bytes, _ := tc.serializer.Encode(&record)
+			newRecord := &analytics.AnalyticsRecord{}
+
+			tc.serializer.Decode(bytes, newRecord)
```
Comment (matiasinsaurralde) [inline serializer/serializer_test.go:65]: Error check, just in case!

**Review by matiasinsaurralde — COMMENTED** [review]
Added some initial comments

File: serializer/serializer_test.go (hunk 2 above)
Comment (sredxny) [reply serializer/serializer_test.go]: done

File: main.go (hunk 1 above)
Comment (sredxny) [reply main.go:244]: done

File: pumps/mongo.go [hunk 3]
```diff
@@ -598,8 +598,8 @@ func (m *MongoPump) WriteUptimeData(data []interface{}) {
 		decoded := analytics.UptimeReportData{}
 
 		if err := msgpack.Unmarshal([]byte(v.(string)), &decoded); err != nil {
+			// ToDo: should this work with serializer?
```
Comment (sredxny) [inline pumps/mongo.go:601]: @matiasinsaurralde @tbuchaillot

File: pumps/mongo_selective.go [hunk 4]
```diff
@@ -291,11 +291,11 @@ func (m *MongoSelectivePump) WriteUptimeData(data []interface{}) {
 
 			for i, v := range data {
 				decoded := analytics.UptimeReportData{}
+				// ToDo: should this work with serializer?
```
Comment (sredxny) [inline pumps/mongo_selective.go:294]: @matiasinsaurralde @tbuchaillot

File: pumps/sql.go [hunk 5]
```diff
@@ -254,6 +254,7 @@ func (c *SQLPump) WriteUptimeData(data []interface{}) {
 	for i, v := range data {
 		decoded := analytics.UptimeReportData{}
 		if err := msgpack.Unmarshal([]byte(v.(string)), &decoded); err != nil {
+			// ToDo: should this work with serializer?
```
Comment (sredxny) [inline pumps/sql.go:257]: @tbuchaillot @matiasinsaurralde

File: serializer/go_tiny.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package serializer
```
Comment (tbuchaillot) [inline serializer/go_tiny.go:1]: we need to remove this

File: serializer/protobuf.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	return transformedRecs
+}
+
+func (pb *ProtobufSerializer) TransfromSingleRecordToProto(rec analytics.AnalyticsRecord) analyticsproto.AnalyticsRecord {
+
+	newRec := analyticsproto.AnalyticsRecord{}
+	copier.Copy(&newRec, &rec)
+
+	// get original huso horario
+	// grab the ms
```
Comment (tbuchaillot) [inline serializer/protobuf.go:50]: remove the comments

File: serializer/protobuf.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	for i, _ := range recs {
+		transformedRecs[i] = pb.TransfromSingleRecordToProto(recs[i])
+	}
+	return transformedRecs
+}
+
+func (pb *ProtobufSerializer) TransfromSingleRecordToProto(rec analytics.AnalyticsRecord) analyticsproto.AnalyticsRecord {
+
+	newRec := analyticsproto.AnalyticsRecord{}
+	copier.Copy(&newRec, &rec)
```
Comment (tbuchaillot) [inline serializer/protobuf.go:47]: we are relying too much on `copier.Copy`. It seems necessary to create a function that assigns field by field and avoid reflection used on that lib.

File: serializer/protobuf.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	err := copier.Copy(&record, protoRecord)
+	TimeStampFromProto(protoRecord, record)
+
+	return err
+}
+
+// TimestampToProto will process timestamps and assign them to the proto record
+// protobuf converts all timestamps to UTC so we need to ensure that we keep
+// the same original location, in order to do so, we store the location
+func TimestampToProto(newRecord *analyticsproto.AnalyticsRecord, record analytics.AnalyticsRecord) {
```
Comment (tbuchaillot) [inline serializer/protobuf.go:70]: Wouldn't be better to have this as a helper function for `analyticsproto.AnalyticsRecord`?

File: serializer/serializer.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	GetSuffix() string
+}
+
+const MSGP_SERIALIZER = "msgpack"
+const GOTINY_SERIALIZER = "gotiny"
+const PROTOBUF_SERIALIZER = "protobuf"
+
+func NewAnalyticsSerializer(serializerType string) AnalyticsSerializer {
+	switch serializerType {
+	case GOTINY_SERIALIZER:
```
Comment (tbuchaillot) [inline serializer/serializer.go:25]: remove this

File: pumps/mongo.go (hunk 3 above)
Comment (tbuchaillot) [reply pumps/mongo.go:601]: Good point @sredxny. Probably is better to have consistency between uptime analytics and common analytics. So yeah, we need to implement something at pump and Gateway level for uptime analytics.

**Review by tbuchaillot — CHANGES_REQUESTED** [review]

File: serializer/go_tiny.go (hunk 6 above)
Comment (sredxny) [reply serializer/go_tiny.go]: done

File: serializer/protobuf.go (hunk 7 above)
Comment (sredxny) [reply serializer/protobuf.go]: done

File: serializer/protobuf.go (hunk 8 above)
Comment (sredxny) [reply serializer/protobuf.go]: done

File: serializer/serializer.go (hunk 10 above)
Comment (sredxny) [reply serializer/serializer.go]: done

File: serializer/protobuf.go (hunk 9 above)
Comment (matiasinsaurralde) [reply serializer/protobuf.go:122]: Agree

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Review by matiasinsaurralde — APPROVED** [review]

**Pull request merged**